File: src/main/java/com/fluxchess/pulse/Notation.java
/*
 * Copyright (C) 2013-2015 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */
package com.fluxchess.pulse;

import com.fluxchess.jcpi.models.GenericBoard;
import com.fluxchess.jcpi.models.GenericCastling;
import com.fluxchess.jcpi.models.GenericChessman;
import com.fluxchess.jcpi.models.GenericColor;
import com.fluxchess.jcpi.models.GenericFile;
import com.fluxchess.jcpi.models.GenericPiece;
import com.fluxchess.jcpi.models.GenericPosition;
import com.fluxchess.jcpi.models.GenericRank;
import com.fluxchess.jcpi.models.IllegalNotationException;
import org.jetbrains.annotations.NotNull;

import static com.fluxchess.pulse.Castling.BLACK_KINGSIDE;
import static com.fluxchess.pulse.Castling.BLACK_QUEENSIDE;
import static com.fluxchess.pulse.Castling.NOCASTLING;
import static com.fluxchess.pulse.Castling.WHITE_KINGSIDE;
import static com.fluxchess.pulse.Castling.WHITE_QUEENSIDE;
import static com.fluxchess.pulse.CastlingType.KINGSIDE;
import static com.fluxchess.pulse.CastlingType.NOCASTLINGTYPE;
import static com.fluxchess.pulse.CastlingType.QUEENSIDE;
import static com.fluxchess.pulse.Color.BLACK;
import static com.fluxchess.pulse.Color.NOCOLOR;
import static com.fluxchess.pulse.Color.WHITE;
import static com.fluxchess.pulse.File.NOFILE;
import static com.fluxchess.pulse.File.a;
import static com.fluxchess.pulse.File.b;
import static com.fluxchess.pulse.File.c;
import static com.fluxchess.pulse.File.d;
import static com.fluxchess.pulse.File.e;
import static com.fluxchess.pulse.File.f;
import static com.fluxchess.pulse.File.g;
import static com.fluxchess.pulse.File.h;
import static com.fluxchess.pulse.Piece.BLACK_BISHOP;
import static com.fluxchess.pulse.Piece.BLACK_KING;
import static com.fluxchess.pulse.Piece.BLACK_KNIGHT;
import static com.fluxchess.pulse.Piece.BLACK_PAWN;
import static com.fluxchess.pulse.Piece.BLACK_QUEEN;
import static com.fluxchess.pulse.Piece.BLACK_ROOK;
import static com.fluxchess.pulse.Piece.NOPIECE;
import static com.fluxchess.pulse.Piece.WHITE_BISHOP;
import static com.fluxchess.pulse.Piece.WHITE_KING;
import static com.fluxchess.pulse.Piece.WHITE_KNIGHT;
import static com.fluxchess.pulse.Piece.WHITE_PAWN;
import static com.fluxchess.pulse.Piece.WHITE_QUEEN;
import static com.fluxchess.pulse.Piece.WHITE_ROOK;
import static com.fluxchess.pulse.PieceType.BISHOP;
import static com.fluxchess.pulse.PieceType.KING;
import static com.fluxchess.pulse.PieceType.KNIGHT;
import static com.fluxchess.pulse.PieceType.NOPIECETYPE;
import static com.fluxchess.pulse.PieceType.PAWN;
import static com.fluxchess.pulse.PieceType.QUEEN;
import static com.fluxchess.pulse.PieceType.ROOK;
import static com.fluxchess.pulse.Rank.NORANK;
import static com.fluxchess.pulse.Rank.r1;
import static com.fluxchess.pulse.Rank.r2;
import static com.fluxchess.pulse.Rank.r3;
import static com.fluxchess.pulse.Rank.r4;
import static com.fluxchess.pulse.Rank.r5;
import static com.fluxchess.pulse.Rank.r6;
import static com.fluxchess.pulse.Rank.r7;
import static com.fluxchess.pulse.Rank.r8;
import static com.fluxchess.pulse.Square.NOSQUARE;

final class Notation {

  static final String STANDARDPOSITION = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1";

  private Notation() {
  }

  static Position toPosition(String fen) {
    try {
      return toPosition(new GenericBoard(fen));
    } catch (IllegalNotationException e) {
      throw new IllegalArgumentException(e);
    }
  }

  static String fromPosition(Position position) {
    return toGenericBoard(position).toString();
  }

  static Position toPosition(@NotNull GenericBoard genericBoard) {
    Position newPosition = new Position();

    // Initialize board
    for (int square : Square.values) {
      GenericPiece genericPiece = genericBoard.getPiece(fromSquare(square));
      if (genericPiece != null) {
        int piece = toPiece(genericPiece);
        newPosition.put(piece, square);
      }
    }

    // Initialize active color
    newPosition.setActiveColor(toColor(genericBoard.getActiveColor()));

    // Initialize castling
    for (int color : Color.values) {
      for (int castlingtype : CastlingType.values) {
        GenericFile genericFile = genericBoard.getCastling(
            fromColor(color), fromCastlingType(castlingtype)
        );
        if (genericFile != null) {
          newPosition.setCastlingRight(Castling.valueOf(color, castlingtype));
        }
      }
    }

    // Initialize en passant
    if (genericBoard.getEnPassant() != null) {
      newPosition.setEnPassantSquare(toSquare(genericBoard.getEnPassant()));
    }

    // Initialize half move clock
    newPosition.setHalfmoveClock(genericBoard.getHalfMoveClock());

    // Initialize the full move number
    newPosition.setFullmoveNumber(genericBoard.getFullMoveNumber());

    return newPosition;
  }

  static GenericBoard toGenericBoard(Position position) {
    GenericBoard genericBoard = new GenericBoard();

    // Set board
    for (int square : Square.values) {
      if (position.board[square] != NOPIECE) {
        genericBoard.setPiece(fromPiece(position.board[square]), fromSquare(square));
      }
    }

    // Set castling
    if ((position.castlingRights & WHITE_KINGSIDE) != NOCASTLING) {
      genericBoard.setCastling(
          fromColor(WHITE), fromCastlingType(KINGSIDE),
          fromFile(h)
      );
    }
    if ((position.castlingRights & WHITE_QUEENSIDE) != NOCASTLING) {
      genericBoard.setCastling(
          fromColor(WHITE), fromCastlingType(QUEENSIDE),
          fromFile(a)
      );
    }
    if ((position.castlingRights & BLACK_KINGSIDE) != NOCASTLING) {
      genericBoard.setCastling(
          fromColor(BLACK), fromCastlingType(KINGSIDE),
          fromFile(h)
      );
    }
    if ((position.castlingRights & BLACK_QUEENSIDE) != NOCASTLING) {
      genericBoard.setCastling(
          fromColor(BLACK), fromCastlingType(QUEENSIDE),
          fromFile(a)
      );
    }

    // Set en passant
    if (position.enPassantSquare != NOSQUARE) {
      genericBoard.setEnPassant(fromSquare(position.enPassantSquare));
    }

    // Set active color
    genericBoard.setActiveColor(fromColor(position.activeColor));

    // Set half move clock
    genericBoard.setHalfMoveClock(position.halfmoveClock);

    // Set full move number
    genericBoard.setFullMoveNumber(position.getFullmoveNumber());

    return genericBoard;
  }

  static int toColor(@NotNull GenericColor genericColor) {
    switch (genericColor) {
      case WHITE:
        return WHITE;
      case BLACK:
        return BLACK;
      default:
        throw new IllegalArgumentException();
    }
  }

  static GenericColor fromColor(int color) {
    switch (color) {
      case WHITE:
        return GenericColor.WHITE;
      case BLACK:
        return GenericColor.BLACK;
      case NOCOLOR:
      default:
        throw new IllegalArgumentException();
    }
  }

  static GenericChessman fromPieceType(int piecetype) {
    switch (piecetype) {
      case PAWN:
        return GenericChessman.PAWN;
      case KNIGHT:
        return GenericChessman.KNIGHT;
      case BISHOP:
        return GenericChessman.BISHOP;
      case ROOK:
        return GenericChessman.ROOK;
      case QUEEN:
        return GenericChessman.QUEEN;
      case KING:
        return GenericChessman.KING;
      case NOPIECETYPE:
      default:
        throw new IllegalArgumentException();
    }
  }

  static int toPiece(@NotNull GenericPiece genericPiece) {
    switch (genericPiece) {
      case WHITEPAWN:
        return WHITE_PAWN;
      case WHITEKNIGHT:
        return WHITE_KNIGHT;
      case WHITEBISHOP:
        return WHITE_BISHOP;
      case WHITEROOK:
        return WHITE_ROOK;
      case WHITEQUEEN:
        return WHITE_QUEEN;
      case WHITEKING:
        return WHITE_KING;
      case BLACKPAWN:
        return BLACK_PAWN;
      case BLACKKNIGHT:
        return BLACK_KNIGHT;
      case BLACKBISHOP:
        return BLACK_BISHOP;
      case BLACKROOK:
        return BLACK_ROOK;
      case BLACKQUEEN:
        return BLACK_QUEEN;
      case BLACKKING:
        return BLACK_KING;
      default:
        throw new IllegalArgumentException();
    }
  }

  static GenericPiece fromPiece(int piece) {
    switch (piece) {
      case WHITE_PAWN:
        return GenericPiece.WHITEPAWN;
      case WHITE_KNIGHT:
        return GenericPiece.WHITEKNIGHT;
      case WHITE_BISHOP:
        return GenericPiece.WHITEBISHOP;
      case WHITE_ROOK:
        return GenericPiece.WHITEROOK;
      case WHITE_QUEEN:
        return GenericPiece.WHITEQUEEN;
      case WHITE_KING:
        return GenericPiece.WHITEKING;
      case BLACK_PAWN:
        return GenericPiece.BLACKPAWN;
      case BLACK_KNIGHT:
        return GenericPiece.BLACKKNIGHT;
      case BLACK_BISHOP:
        return GenericPiece.BLACKBISHOP;
      case BLACK_ROOK:
        return GenericPiece.BLACKROOK;
      case BLACK_QUEEN:
        return GenericPiece.BLACKQUEEN;
      case BLACK_KING:
        return GenericPiece.BLACKKING;
      case NOPIECE:
      default:
        throw new IllegalArgumentException();
    }
  }

  static GenericCastling fromCastlingType(int castlingtype) {
    switch (castlingtype) {
      case KINGSIDE:
        return GenericCastling.KINGSIDE;
      case QUEENSIDE:
        return GenericCastling.QUEENSIDE;
      case NOCASTLINGTYPE:
      default:
        throw new IllegalArgumentException();
    }
  }

  static int toFile(@NotNull GenericFile genericFile) {
    switch (genericFile) {
      case Fa:
        return a;
      case Fb:
        return b;
      case Fc:
        return c;
      case Fd:
        return d;
      case Fe:
        return e;
      case Ff:
        return f;
      case Fg:
        return g;
      case Fh:
        return h;
      default:
        throw new IllegalArgumentException();
    }
  }

  static GenericFile fromFile(int file) {
    switch (file) {
      case a:
        return GenericFile.Fa;
      case b:
        return GenericFile.Fb;
      case c:
        return GenericFile.Fc;
      case d:
        return GenericFile.Fd;
      case e:
        return GenericFile.Fe;
      case f:
        return GenericFile.Ff;
      case g:
        return GenericFile.Fg;
      case h:
        return GenericFile.Fh;
      case NOFILE:
      default:
        throw new IllegalArgumentException();
    }
  }

  static int toRank(@NotNull GenericRank genericRank) {
    switch (genericRank) {
      case R1:
        return r1;
      case R2:
        return r2;
      case R3:
        return r3;
      case R4:
        return r4;
      case R5:
        return r5;
      case R6:
        return r6;
      case R7:
        return r7;
      case R8:
        return r8;
      default:
        throw new IllegalArgumentException();
    }
  }

  static GenericRank fromRank(int rank) {
    switch (rank) {
      case r1:
        return GenericRank.R1;
      case r2:
        return GenericRank.R2;
      case r3:
        return GenericRank.R3;
      case r4:
        return GenericRank.R4;
      case r5:
        return GenericRank.R5;
      case r6:
        return GenericRank.R6;
      case r7:
        return GenericRank.R7;
      case r8:
        return GenericRank.R8;
      case NORANK:
      default:
        throw new IllegalArgumentException();
    }
  }

  static int toSquare(@NotNull GenericPosition genericPosition) {
    int square = toRank(genericPosition.rank) * 16 + toFile(genericPosition.file);
    assert Square.isValid(square);

    return square;
  }

  static GenericPosition fromSquare(int square) {
    assert Square.isValid(square);

    return GenericPosition.valueOf(fromFile(Square.getFile(square)), fromRank(Square.getRank(square)));
  }
}

File: pulse-kotlin/src/commonMain/kotlin/com/fluxchess/pulse/kotlin/uci/Notation.kt
/*
 * Copyright 2013-2025 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package com.fluxchess.pulse.kotlin.uci

import com.fluxchess.pulse.kotlin.engine.BISHOP
import com.fluxchess.pulse.kotlin.engine.BLACK
import com.fluxchess.pulse.kotlin.engine.BLACK_BISHOP
import com.fluxchess.pulse.kotlin.engine.BLACK_KING
import com.fluxchess.pulse.kotlin.engine.BLACK_KINGSIDE
import com.fluxchess.pulse.kotlin.engine.BLACK_KNIGHT
import com.fluxchess.pulse.kotlin.engine.BLACK_PAWN
import com.fluxchess.pulse.kotlin.engine.BLACK_QUEEN
import com.fluxchess.pulse.kotlin.engine.BLACK_QUEENSIDE
import com.fluxchess.pulse.kotlin.engine.BLACK_ROOK
import com.fluxchess.pulse.kotlin.engine.Castling
import com.fluxchess.pulse.kotlin.engine.CastlingType
import com.fluxchess.pulse.kotlin.engine.Color
import com.fluxchess.pulse.kotlin.engine.FILE_A
import com.fluxchess.pulse.kotlin.engine.FILE_B
import com.fluxchess.pulse.kotlin.engine.FILE_C
import com.fluxchess.pulse.kotlin.engine.FILE_D
import com.fluxchess.pulse.kotlin.engine.FILE_E
import com.fluxchess.pulse.kotlin.engine.FILE_F
import com.fluxchess.pulse.kotlin.engine.FILE_G
import com.fluxchess.pulse.kotlin.engine.FILE_H
import com.fluxchess.pulse.kotlin.engine.File
import com.fluxchess.pulse.kotlin.engine.KING
import com.fluxchess.pulse.kotlin.engine.KINGSIDE
import com.fluxchess.pulse.kotlin.engine.KNIGHT
import com.fluxchess.pulse.kotlin.engine.Move
import com.fluxchess.pulse.kotlin.engine.NO_FILE
import com.fluxchess.pulse.kotlin.engine.NO_PIECE
import com.fluxchess.pulse.kotlin.engine.NO_PIECE_TYPE
import com.fluxchess.pulse.kotlin.engine.NO_SQUARE
import com.fluxchess.pulse.kotlin.engine.PAWN
import com.fluxchess.pulse.kotlin.engine.Piece
import com.fluxchess.pulse.kotlin.engine.PieceType
import com.fluxchess.pulse.kotlin.engine.Position
import com.fluxchess.pulse.kotlin.engine.QUEEN
import com.fluxchess.pulse.kotlin.engine.QUEENSIDE
import com.fluxchess.pulse.kotlin.engine.RANK_1
import com.fluxchess.pulse.kotlin.engine.RANK_2
import com.fluxchess.pulse.kotlin.engine.RANK_3
import com.fluxchess.pulse.kotlin.engine.RANK_4
import com.fluxchess.pulse.kotlin.engine.RANK_5
import com.fluxchess.pulse.kotlin.engine.RANK_6
import com.fluxchess.pulse.kotlin.engine.RANK_7
import com.fluxchess.pulse.kotlin.engine.RANK_8
import com.fluxchess.pulse.kotlin.engine.ROOK
import com.fluxchess.pulse.kotlin.engine.Rank
import com.fluxchess.pulse.kotlin.engine.Square
import com.fluxchess.pulse.kotlin.engine.WHITE
import com.fluxchess.pulse.kotlin.engine.WHITE_BISHOP
import com.fluxchess.pulse.kotlin.engine.WHITE_KING
import com.fluxchess.pulse.kotlin.engine.WHITE_KINGSIDE
import com.fluxchess.pulse.kotlin.engine.WHITE_KNIGHT
import com.fluxchess.pulse.kotlin.engine.WHITE_PAWN
import com.fluxchess.pulse.kotlin.engine.WHITE_QUEEN
import com.fluxchess.pulse.kotlin.engine.WHITE_QUEENSIDE
import com.fluxchess.pulse.kotlin.engine.WHITE_ROOK
import com.fluxchess.pulse.kotlin.engine.castlingOf
import com.fluxchess.pulse.kotlin.engine.fileOf
import com.fluxchess.pulse.kotlin.engine.files
import com.fluxchess.pulse.kotlin.engine.isValidFile
import com.fluxchess.pulse.kotlin.engine.originSquareOf
import com.fluxchess.pulse.kotlin.engine.pieceOf
import com.fluxchess.pulse.kotlin.engine.promotionOf
import com.fluxchess.pulse.kotlin.engine.rankOf
import com.fluxchess.pulse.kotlin.engine.ranks
import com.fluxchess.pulse.kotlin.engine.squareOf
import com.fluxchess.pulse.kotlin.engine.targetSquareOf

private val fenTokenRegex = "\\s+".toRegex()
private val rankRegex = "/".toRegex()

fun String.toPositionOrNull(): Position? = try {
	toPosition()
} catch (e: FenException) {
	null
}

fun String.toPosition(): Position {
	// Clean and split into tokens
	val tokens = trim().split(fenTokenRegex)

	// halfmove clock and fullmove number are optional
	if (tokens.size < 4 || tokens.size > 6) {
		throw FenException("Invalid FEN: $this")
	}

	val position = Position()

	// Parse board
	val board = tokens[0].split(rankRegex)
	if (board.size != 8) {
		throw FenException("Invalid board: ${tokens[0]}")
	}
	var currentRank = RANK_8
	board.forEach { rank ->
		var currentFile = FILE_A
		rank.forEach { char ->
			if (!isValidFile(currentFile)) {
				throw FenException("Invalid rank: $rank")
			}
			val piece = char.toPieceOrNull()
			if (piece != null) {
				position.put(piece, squareOf(currentFile, currentRank))
				if (currentFile == FILE_H) {
					currentFile = NO_FILE
				} else {
					currentFile++
				}
				return@forEach
			}
			val emptySquares = char.toEmptySquaresOrNull()
			if (emptySquares != null) {
				currentFile += emptySquares
				return@forEach
			}
		}
		if (currentFile != NO_FILE) {
			throw FenException("Invalid rank: $rank")
		}
		currentRank--
	}

	// Parse active color
	val activeColor = tokens[1].toActiveColor()
	position.activeColor = activeColor

	// Parse castling rights
	if (tokens[2] != "-") {
		if (tokens[2].length > 4) {
			throw FenException("Invalid castling rights: ${tokens[2]}")
		}
		tokens[2].forEach { char ->
			val castling = char.toCastling()
			position.setCastlingRight(castling)
		}
	}

	// Parse en passant square
	if (tokens[3] != "-") {
		if (tokens[3].length != 2) {
			throw FenException("Invalid en passant square: ${tokens[3]}")
		}
		val file = tokens[3][0].toFile()
		val rank = tokens[3][1].toRank()
		if (!(activeColor == WHITE && rank == RANK_6) && !(activeColor == BLACK && rank == RANK_3)) {
			throw FenException("Invalid en passant square: ${tokens[3]}")
		}
		position.enPassantSquare = squareOf(file, rank)
	}

	// Parse halfmove clock
	if (tokens.size >= 5) {
		val halfmoveClock = tokens[4].toIntOrNull() ?: throw FenException("Invalid halfmove clock: ${tokens[4]}")
		position.halfmoveClock = halfmoveClock
	}

	// Parse fullmove number
	if (tokens.size == 6) {
		val fullmoveNumber = tokens[5].toIntOrNull() ?: throw FenException("Invalid fullmove number: ${tokens[5]}")
		position.halfmoveNumber = fullmoveNumber * 2
		if (activeColor == BLACK) {
			position.halfmoveNumber++
		}
	} else {
		position.halfmoveNumber = 2
	}

	return position
}

class FenException(message: String) : Exception(message)

private fun Char.toPieceOrNull(): Piece? = toPieceTypeOrNull()?.let { pieceOf(toColor(), it) }

private fun Char.toColor(): Color = if (isLowerCase()) BLACK else WHITE

private fun Char.toPieceTypeOrNull(): PieceType? = when (lowercaseChar()) {
	'p' -> PAWN
	'n' -> KNIGHT
	'b' -> BISHOP
	'r' -> ROOK
	'q' -> QUEEN
	'k' -> KING
	else -> null
}

private fun Char.toEmptySquaresOrNull(): Int? {
	val emptySquares = if (isDigit()) digitToInt() else return null
	return if (emptySquares in 1..8) emptySquares else null
}

private fun String.toActiveColor(): Color = when (lowercase()) {
	"w" -> WHITE
	"b" -> BLACK
	else -> throw FenException("Invalid active color: $this")
}

private fun Char.toCastling(): Castling = castlingOf(toColor(), toCastlingType())

private fun Char.toCastlingType(): CastlingType = when (lowercaseChar()) {
	'k' -> KINGSIDE
	'q' -> QUEENSIDE
	else -> throw FenException("Invalid castling type: $this")
}

private fun Char.toFile(): File = when (lowercaseChar()) {
	'a' -> FILE_A
	'b' -> FILE_B
	'c' -> FILE_C
	'd' -> FILE_D
	'e' -> FILE_E
	'f' -> FILE_F
	'g' -> FILE_G
	'h' -> FILE_H
	else -> throw FenException("Invalid file: $this")
}

private fun Char.toRank(): Rank = when (lowercaseChar()) {
	'1' -> RANK_1
	'2' -> RANK_2
	'3' -> RANK_3
	'4' -> RANK_4
	'5' -> RANK_5
	'6' -> RANK_6
	'7' -> RANK_7
	'8' -> RANK_8
	else -> throw FenException("Invalid rank: $this")
}

fun Position.toFEN(): String {
	var fen = ""

	// board
	for (rank in ranks.reversed()) {
		var emptySquares = 0
		for (file in files) {
			val piece = get(squareOf(file, rank))
			if (piece == NO_PIECE) {
				emptySquares++
			} else {
				if (emptySquares > 0) {
					fen += emptySquares
					emptySquares = 0
				}
				fen += pieceToNotation(piece)
			}
		}
		if (emptySquares > 0) {
			fen += emptySquares
		}
		if (rank > RANK_1) {
			fen += '/'
		}
	}

	// active color
	fen += " " + colorToNotation(activeColor)

	// castling rights
	var rights = ""
	if (castlingRights and WHITE_KINGSIDE == WHITE_KINGSIDE) {
		rights += "K"
	}
	if (castlingRights and WHITE_QUEENSIDE == WHITE_QUEENSIDE) {
		rights += "Q"
	}
	if (castlingRights and BLACK_KINGSIDE == BLACK_KINGSIDE) {
		rights += "k"
	}
	if (castlingRights and BLACK_QUEENSIDE == BLACK_QUEENSIDE) {
		rights += "q"
	}
	if (rights == "") {
		rights = "-"
	}
	fen += " $rights"

	// en passant square
	if (enPassantSquare != NO_SQUARE) {
		fen += " " + squareToNotation(enPassantSquare)
	} else {
		fen += " -"
	}

	// halfmove clock
	fen += " $halfmoveClock"

	// fullmove number
	fen += " " + (halfmoveNumber / 2)

	return fen
}

private fun pieceToNotation(piece: Piece): String = when (piece) {
	WHITE_PAWN -> "P"
	WHITE_KNIGHT -> "N"
	WHITE_BISHOP -> "B"
	WHITE_ROOK -> "R"
	WHITE_QUEEN -> "Q"
	WHITE_KING -> "K"
	BLACK_PAWN -> "p"
	BLACK_KNIGHT -> "n"
	BLACK_BISHOP -> "b"
	BLACK_ROOK -> "r"
	BLACK_QUEEN -> "q"
	BLACK_KING -> "k"
	else -> error("Invalid piece: $piece")
}

private fun colorToNotation(color: Color): String = when (color) {
	WHITE -> "w"
	BLACK -> "b"
	else -> error("Invalid color: $color")
}

private fun squareToNotation(square: Square): String {
	return fileToNotation(fileOf(square)) + rankToNotation(rankOf(square))
}

private fun fileToNotation(file: File): String = when (file) {
	FILE_A -> "a"
	FILE_B -> "b"
	FILE_C -> "c"
	FILE_D -> "d"
	FILE_E -> "e"
	FILE_F -> "f"
	FILE_G -> "g"
	FILE_H -> "h"
	else -> error("Invalid file: $file")
}

private fun rankToNotation(rank: Rank): String = when (rank) {
	RANK_1 -> "1"
	RANK_2 -> "2"
	RANK_3 -> "3"
	RANK_4 -> "4"
	RANK_5 -> "5"
	RANK_6 -> "6"
	RANK_7 -> "7"
	RANK_8 -> "8"
	else -> error("Invalid rank: $rank")
}

fun Move.toNotation(): String {
	var notation = ""
	notation += squareToNotation(originSquareOf(this))
	notation += squareToNotation(targetSquareOf(this))
	val promotion = promotionOf(this)
	if (promotion != NO_PIECE_TYPE) {
		notation += pieceTypeToNotation(promotion)
	}
	return notation
}

private fun pieceTypeToNotation(pieceType: PieceType): String = when (pieceType) {
	KNIGHT -> "n"
	BISHOP -> "b"
	ROOK -> "r"
	QUEEN -> "q"
	else -> error("Invalid piece type: $pieceType")
}
